File: omrchecker-js/packages/core/src/processors/detection/bubbles_threshold/interpretationPass.ts
/**
 * Bubbles threshold interpretation pass.
 *
 * TypeScript port of src/processors/detection/bubbles_threshold/interpretation_pass.py
 * Extends FieldTypeInterpretationPass for bubble field interpretation.
 */

import { Logger } from '../../../utils/logger';
import { FieldTypeInterpretationPass } from '../base/interpretationPass';
import type { Field } from '../../layout/field/base';
import { NumberAggregate } from '../../../utils/stats';
import { BubblesFieldInterpretation } from './interpretation';
import { GlobalThreshold, type ThresholdConfig } from '../../threshold/GlobalThreshold';
import type { TuningConfig, FieldLevelAggregates, FileLevelAggregates } from '../base/commonPass';
import type { BubbleMeanValue } from '../models/detectionResults';
import { DetectionRepository } from '../../repositories/DetectionRepository';
import type {
  OutlierDeviationThresholdConfig,
  FallbackThresholdConfig,
} from '../../../schemas/models/config';

const logger = new Logger('BubblesThresholdInterpretationPass');

/**
 * Interpretation pass for bubble fields.
 *
 * Manages file-level aggregates for threshold calculation and interpretation.
 * Uses BubblesFieldInterpretation for actual interpretation.
 */
export class BubblesThresholdInterpretationPass extends FieldTypeInterpretationPass {
  private repository: DetectionRepository;

  constructor(
    tuningConfig: TuningConfig,
    fieldDetectionType: string,
    repository: DetectionRepository
  ) {
    super(tuningConfig, fieldDetectionType);
    this.repository = repository;
  }

  /**
   * Get field interpretation for a bubble field.
   *
   * @param field - Field to interpret
   * @param fileLevelDetectionAggregates - File-level detection aggregates
   * @param fileLevelInterpretationAggregates - File-level interpretation aggregates
   * @returns BubblesFieldInterpretation result
   */
  getFieldInterpretation(
    field: Field,
    fileLevelDetectionAggregates: unknown,
    fileLevelInterpretationAggregates: unknown
  ): BubblesFieldInterpretation {
    return new BubblesFieldInterpretation(
      this.tuningConfig,
      field,
      fileLevelDetectionAggregates,
      fileLevelInterpretationAggregates
    );
  }

  /**
   * Initialize file-level aggregates for interpretation.
   *
   * @param filePath - Path to the file being processed
   */
  initializeFileLevelAggregates(filePath: string): void {
    super.initializeFileLevelAggregates(filePath);

    // Get bubble means from repository
    const allBubbleMeans = this.repository.getAllBubbleMeansForCurrentFile();

    // Calculate std deviations from repository results
    const allOutlierDeviations: number[] = [];
    for (const fieldResult of this.repository.getAllBubbleFieldsForCurrentFile().values()) {
      allOutlierDeviations.push(fieldResult.stdDeviation);
    }

    const fieldWiseMeansAndRefs = allBubbleMeans;
    const outlierDeviationThresholdForFile =
      this.getOutlierDeviationThreshold(allOutlierDeviations);
    const { fileLevelFallbackThreshold, globalMaxJump } =
      this.getFallbackThreshold(fieldWiseMeansAndRefs);

    logger.debug(
      `Thresholding: file_level_fallback_threshold: ${fileLevelFallbackThreshold.toFixed(2)} global_std_THR: ${outlierDeviationThresholdForFile.toFixed(2)} ${
        fileLevelFallbackThreshold === 255 ? '(Looks like a Xeroxed OMR)' : ''
      }`
    );

    this.insertFileLevelAggregates({
      file_level_fallback_threshold: fileLevelFallbackThreshold,
      global_max_jump: globalMaxJump,
      outlier_deviation_threshold_for_file: outlierDeviationThresholdForFile,
      field_label_wise_local_thresholds: {},
      bubble_field_type_wise_thresholds: {},
      all_fields_local_thresholds: new NumberAggregate(),
      field_wise_confidence_metrics: {},
    });
  }

  /**
   * Get outlier deviation threshold.
   *
   * @param allOutlierDeviations - List of standard deviations
   * @returns Outlier deviation threshold
   */
  getOutlierDeviationThreshold(allOutlierDeviations: number[]): number {
    const config = this.tuningConfig.thresholding as OutlierDeviationThresholdConfig;

    const minJumpStd = config.min_jump_std || 5.0;
    const globalPageThresholdStd = config.global_page_threshold_std || 10.0;

    // Use GlobalThresholdStrategy
    const strategy = new GlobalThreshold();
    const thresholdConfig: ThresholdConfig = {
      minJump: minJumpStd,
      defaultThreshold: globalPageThresholdStd,
    };

    // allOutlierDeviations is already a list of floats (std deviations)
    const result = strategy.calculateThreshold(allOutlierDeviations, thresholdConfig);
    return result.thresholdValue;
  }

  /**
   * Get fallback threshold.
   *
   * @param fieldWiseMeansAndRefs - List of BubbleMeanValue objects
   * @returns Object with fileLevelFallbackThreshold and globalMaxJump
   */
  getFallbackThreshold(fieldWiseMeansAndRefs: BubbleMeanValue[]): {
    fileLevelFallbackThreshold: number;
    globalMaxJump: number;
  } {
    const config = this.tuningConfig.thresholding as FallbackThresholdConfig;

    const globalPageThreshold = config.global_page_threshold || 180;
    const minJump = config.min_jump || 10;

    // Use GlobalThresholdStrategy
    const strategy = new GlobalThreshold();
    const thresholdConfig: ThresholdConfig = {
      minJump,
      defaultThreshold: globalPageThreshold,
    };

    // fieldWiseMeansAndRefs is a list of BubbleMeanValue objects
    const bubbleValues = fieldWiseMeansAndRefs.map((item) => item.meanValue);

    const result = strategy.calculateThreshold(bubbleValues, thresholdConfig);

    // Approximate global_max_jump from result
    const globalMaxJump = result.maxJump;

    return {
      fileLevelFallbackThreshold: result.thresholdValue,
      globalMaxJump,
    };
  }

  /**
   * Update field-level aggregates after interpretation.
   *
   * @param field - Field that was processed
   * @param fieldInterpretation - Interpretation result
   */
  updateFieldLevelAggregatesOnProcessedFieldInterpretation(
    field: Field,
    fieldInterpretation: BubblesFieldInterpretation
  ): void {
    super.updateFieldLevelAggregatesOnProcessedFieldInterpretation(
      field,
      fieldInterpretation
    );

    this.insertFieldLevelAggregates({
      is_multi_marked: fieldInterpretation.isMultiMarked,
      local_threshold_for_field: fieldInterpretation.localThresholdForField,
      bubble_interpretations: fieldInterpretation.bubbleInterpretations,
    });
  }

  /**
   * Update file-level aggregates after field interpretation.
   *
   * @param field - Field that was processed
   * @param fieldInterpretation - Interpretation result
   * @param fieldLevelAggregates - Field-level aggregates
   */
  updateFileLevelAggregatesOnProcessedFieldInterpretation(
    field: Field,
    fieldInterpretation: BubblesFieldInterpretation,
    fieldLevelAggregates: FieldLevelAggregates
  ): void {
    super.updateFileLevelAggregatesOnProcessedFieldInterpretation(
      field,
      fieldInterpretation,
      fieldLevelAggregates
    );

    const fileAgg = this.getFileLevelAggregates();
    if (!fileAgg) {
      throw new Error('File level aggregates not initialized');
    }

    const allFieldsLocalThresholds = fileAgg.all_fields_local_thresholds as NumberAggregate;
    allFieldsLocalThresholds.push(fieldInterpretation.localThresholdForField, field.id);

    // TODO: update bubble_field_type_wise_thresholds
  }

  /**
   * Get file-level interpretation aggregates.
   * Used by TemplateInterpretationPass to collect aggregates from field type runners.
   *
   * @returns File-level aggregates
   */
  getFileLevelInterpretationAggregates(): FileLevelAggregates | undefined {
    return this.getFileLevelAggregates();
  }
}

File: src/processors/detection/bubbles_threshold/interpretation_pass.py
from src.processors.detection.base.interpretation_pass import (
    FieldTypeInterpretationPass,
)
from src.processors.detection.bubbles_threshold.interpretation import (
    BubblesFieldInterpretation,
)
from src.processors.layout.field.base import Field
from src.processors.detection.detection_repository import DetectionRepository
from src.processors.detection.threshold.global_threshold import GlobalThresholdStrategy
from src.processors.detection.threshold.threshold_result import ThresholdConfig
from src.utils.logger import logger
from src.utils.stats import NumberAggregate


class BubblesThresholdInterpretationPass(FieldTypeInterpretationPass):
    def __init__(self, *args, repository: DetectionRepository, **kwargs) -> None:
        super().__init__(*args, **kwargs)
        self.repository = repository

    # Note: This is used by parent to generate the interpretation: detected string etc
    def get_field_interpretation(
        self,
        field: Field,
        file_level_detection_aggregates,
        file_level_aggregates,
    ) -> BubblesFieldInterpretation:
        tuning_config = self.tuning_config
        return BubblesFieldInterpretation(
            tuning_config,
            field,
            file_level_detection_aggregates,
            file_level_aggregates,
        )

    def initialize_file_level_aggregates(
        self,
        file_path,
    ) -> None:
        super().initialize_file_level_aggregates(file_path)
        # Get bubble means from repository
        all_bubble_means = self.repository.get_all_bubble_means_for_current_file()

        # Calculate std deviations from repository results
        all_outlier_deviations = []
        for (
            field_result
        ) in self.repository.get_all_bubble_fields_for_current_file().values():
            all_outlier_deviations.append(field_result.std_deviation)

        field_wise_means_and_refs = all_bubble_means

        outlier_deviation_threshold_for_file = self.get_outlier_deviation_threshold(
            all_outlier_deviations
        )

        file_level_fallback_threshold, global_max_jump = self.get_fallback_threshold(
            field_wise_means_and_refs
        )

        logger.debug(
            f"Thresholding: \t file_level_fallback_threshold: {round(file_level_fallback_threshold, 2)} \tglobal_std_THR: {round(outlier_deviation_threshold_for_file, 2)}\t{'(Looks like a Xeroxed OMR)' if (file_level_fallback_threshold == 255) else ''}"
        )
        # TODO: uncomment for more granular threshold detection insights (per bubbleFieldType)
        # bubble_field_type_wise_thresholds = {}
        # for field_detection_aggregates in field_label_wise_detection_aggregates.values():
        #     field = field_detection_aggregates["field"]
        #     bubble_field_type = field.bubble_field_type
        #     field_wise_means_and_refs = field_detection_aggregates["field_bubble_means"].values()
        #
        #     if bubble_field_type not in bubble_field_type_wise_thresholds:
        #         bubble_field_type_wise_thresholds[bubble_field_type] = NumberAggregate()

        #     bubble_field_type_level_fallback_threshold, _ = self.get_fallback_threshold(file_path, field_wise_means_and_refs)
        #     bubble_field_type_wise_thresholds[bubble_field_type].push(bubble_field_type_level_fallback_threshold)

        # TODO: return bubble_field_type_wise(MCQ/INT/CUSTOM_1/CUSTOM_2) thresholds too - for interpretation?
        self.insert_file_level_aggregates(
            {
                "file_level_fallback_threshold": file_level_fallback_threshold,
                "global_max_jump": global_max_jump,
                "outlier_deviation_threshold_for_file": outlier_deviation_threshold_for_file,
                "field_label_wise_local_thresholds": {},
                "bubble_field_type_wise_thresholds": {},
                "all_fields_local_thresholds": NumberAggregate(),
                "field_wise_confidence_metrics": {},
            }
        )

    def get_outlier_deviation_threshold(
        self,
        all_outlier_deviations,
    ) -> float:
        """Calculate outlier deviation threshold using threshold strategy.

        Args:
            all_outlier_deviations: List of standard deviations from all fields

        Returns:
            Calculated threshold value
        """
        config = self.tuning_config.thresholding
        threshold_config = ThresholdConfig(
            min_jump=config.min_jump_std,
            default_threshold=config.global_page_threshold_std,
        )

        strategy = GlobalThresholdStrategy()
        result = strategy.calculate_threshold(all_outlier_deviations, threshold_config)
        return result.threshold_value

    def get_fallback_threshold(
        self,
        field_wise_means_and_refs,
    ) -> tuple[float, float]:
        """Calculate fallback threshold using threshold strategy.

        Args:
            field_wise_means_and_refs: List of BubbleMeanValue objects from all fields

        Returns:
            Tuple of (fallback_threshold, global_max_jump)
        """
        config = self.tuning_config.thresholding
        threshold_config = ThresholdConfig(
            min_jump=config.min_jump,
            default_threshold=config.global_page_threshold,
        )

        # field_wise_means_and_refs is a list of BubbleMeanValue objects
        bubble_values = [item.mean_value for item in field_wise_means_and_refs]

        strategy = GlobalThresholdStrategy()
        result = strategy.calculate_threshold(bubble_values, threshold_config)

        # Approximate global_max_jump from result
        global_max_jump = result.max_jump

        return result.threshold_value, global_max_jump

    def update_field_level_aggregates_on_processed_field_interpretation(
        self, field: Field, field_interpretation: BubblesFieldInterpretation
    ) -> None:
        super().update_field_level_aggregates_on_processed_field_interpretation(
            field, field_interpretation
        )
        # TODO: get this object from field_interpretation through a function?
        self.insert_field_level_aggregates(
            {
                # TODO: move is_multi_marked logic to a parent class(or make copy in parent)
                "is_multi_marked": field_interpretation.is_multi_marked,
                "local_threshold_for_field": field_interpretation.local_threshold_for_field,
                "bubble_interpretations": field_interpretation.bubble_interpretations,
                # Needed for exporting?
                # "field_bubble_means": field_interpretation.field_bubble_means,
            }
        )

    def update_file_level_aggregates_on_processed_field_interpretation(
        self, field, field_interpretation, field_level_aggregates
    ) -> None:
        super().update_file_level_aggregates_on_processed_field_interpretation(
            field, field_interpretation, field_level_aggregates
        )
        file_level_aggregates = self.get_file_level_aggregates()
        file_level_aggregates["all_fields_local_thresholds"].push(
            field_interpretation.local_threshold_for_field, field
        )

        # TODO: update bubble_field_type_wise_thresholds
